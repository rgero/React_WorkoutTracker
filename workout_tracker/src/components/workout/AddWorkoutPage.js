import React, { useContext, } from 'react';
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from './WorkoutForm';

export const AddWorkoutPage = ()=> 
{
    const {state, createWorkout} = useContext(WorkoutContext);
    return (
        <Container fluid="md">
            <h1>Create a new workout</h1>
            <WorkoutForm
                errorMessage={state.errorMessage}
                onSubmit={({workout})=> createWorkout({workout})}
            />
        </Container>
    )
}