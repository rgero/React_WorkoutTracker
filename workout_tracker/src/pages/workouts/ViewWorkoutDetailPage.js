import React from 'react';
import {useNavigate, useParams} from "react-router-dom";

import Alert from 'react-bootstrap/Alert';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Button from 'react-bootstrap/Button';
import Card from 'react-bootstrap/Card';
import Dropdown from 'react-bootstrap/Dropdown';

import DateFormatter from '../../helpers/DateFormatter';
import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { ExerciseList } from '../../components/exercise/ExerciseList';

import { confirmAlert } from 'react-confirm-alert'; // Import
import 'react-confirm-alert/src/react-confirm-alert.css'; // Import css

const ViewWorkoutDetailsPage = ()=> 
{
    const navigate = useNavigate();
    const { id } = useParams();

    const {state, fetchWorkouts, deleteWorkout} = React.useContext(WorkoutContext);
    const [currentWorkout, setCurrentWorkout] = React.useState({});
    const [isLoaded, setLoaded] = React.useState(false);
    const [error, setError] = React.useState("");


    const processDelete = async () => {
        confirmAlert({
            title: 'Confirm to delete?',
            message: 'Are you sure to do this.',
            buttons: [
              {
                label: 'Yes',
                onClick: async () => {
                    await deleteWorkout(currentWorkout._id);
                    navigate('/dashboard')
                }
              },
              {
                label: 'No'
              }
            ]
          });

    }

    const isValidWorkout = () => {
        return !Object.keys(currentWorkout).length == 0;
    }

    React.useEffect(()=> {
        const GetWorkoutByID = async (id) => {
            if (!isLoaded)
            {
                await fetchWorkouts();
                setLoaded(true);
            }
            let loadedWorkout = state.filter( (targetWorkout) => {
                return id === targetWorkout._id;
            });

            if (loadedWorkout.length == 0)
            {
                setError("An error has occurred");
            } else {
                setError("");
                setCurrentWorkout(loadedWorkout[0]);
            }
        }
        GetWorkoutByID(id);
    }, [fetchWorkouts, id, isLoaded, state]);

    if (isLoaded) {
        if (isValidWorkout()) { 
            return (
                <Container className="pt-4" fluid="md">
                    <Row className="align-items-center">
                        <Col>
                            <h1>Workout on {DateFormatter(currentWorkout.workoutDate)}</h1> 
                        </Col>
                        <Col>
                            <Dropdown className="float-end">
                                <Dropdown.Toggle id="dropdown-basic">
                                    Options
                                </Dropdown.Toggle>
                                <Dropdown.Menu>
                                    <Dropdown.Item href={`/edit/${currentWorkout._id}`}>Edit</Dropdown.Item>
                                    <Dropdown.Item onClick={processDelete}>Delete</Dropdown.Item>
                                </Dropdown.Menu>
                            </Dropdown>
                        </Col>
                    </Row>
                    <Row>
                        {currentWorkout.notes !== "" ? (
                            <Card>
                                <Card.Body>
                                    <Card.Title>Notes</Card.Title>
                                    <Card.Body>
                                        {currentWorkout.notes}
                                    </Card.Body>
                                </Card.Body>
                            </Card>
                        ) : ( null ) }
                        <Card>
                            <Card.Body>
                                <Card.Title>Exercises - {currentWorkout.exerciseList.length}</Card.Title>
                                <Card.Body>
                                    <ExerciseList exerciseList={currentWorkout.exerciseList}/>
                                </Card.Body>
                                <Card.Body>
                                    
                                </Card.Body>
                            </Card.Body>
                        </Card>
                        <Col className="pt-2"><Button className="float-end" href="/dashboard">Return to Dashboard</Button></Col>
                    </Row>
                </Container>
            )
        } else if (error)
        {
            return(
                <Container>
                    <Alert key="danger" variant="danger">
                        <Alert.Heading>Oh snap! You got an error!</Alert.Heading>
                        <p>{error}</p>
                        <div className="d-flex justify-content-end">
                            <Button onClick={() => navigate('/dashboard')} variant="outline-danger">
                                Return to Dashboard
                            </Button>
                        </div>
                    </Alert>
                    
                </Container>
            )
        }
    } else {
        return (
            <Container>
                Loading
            </Container>
        )
    }
}

export default ViewWorkoutDetailsPage;