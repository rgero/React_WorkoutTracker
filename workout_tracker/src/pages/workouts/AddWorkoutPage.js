import React from 'react';
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from '../../components/workout/WorkoutForm';

export const AddWorkoutPage = ()=> 
{
    const {state, createWorkout} = React.useContext(WorkoutContext);
    return (
        <Container fluid="md">
            <h1>Create a new workout</h1>
            <WorkoutForm
                onSubmit={(workout)=> createWorkout(workout)}
            />
        </Container>
    )
}