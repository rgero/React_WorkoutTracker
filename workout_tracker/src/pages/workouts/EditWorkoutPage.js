import React from 'react';
import {useParams} from "react-router-dom";
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from '../../components/workout/WorkoutForm';

export const EditWorkoutPage = ( )=> 
{
    const {state, fetchWorkouts, updateWorkout} = React.useContext(WorkoutContext);
    const [isLoaded, setLoaded] = React.useState(false);
    const [currentWorkout, setCurrentWorkout] = React.useState({});
    const [error, setError] = React.useState("");
    const { id } = useParams();

    React.useEffect(()=> {
        const GetWorkoutByID = async (id) => {
            if (!isLoaded)
            {
                await fetchWorkouts();
                setLoaded(true);
            }

            let loadedWorkout = state.filter( (targetWorkout) => {
                return id === targetWorkout._id;
            });
                
            if (loadedWorkout.length == 0)
            {
                setError("An error has occurred");
            } else {
                setError("");
                setCurrentWorkout(loadedWorkout[0]);
            }
        }
        GetWorkoutByID(id);
    }, [state]);

    const isValidWorkout = () => {
        return !Object.keys(currentWorkout).length == 0;
    }

    if (isLoaded) {
        if (isValidWorkout()) {
            return (
                <Container fluid="md">
                    <h1>Edit a workout</h1>
                    <WorkoutForm
                        workout={currentWorkout}
                        onSubmit={(workout)=> updateWorkout(workout)}
                    />
                </Container>
                )
        } else if (error)
        {
            return(
                <Container>
                    {error}
                </Container>
            )
        }
    } else {
        return (<div>Loading</div>)
    }
}