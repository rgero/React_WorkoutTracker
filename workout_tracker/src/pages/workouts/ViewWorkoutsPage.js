import React, {useContext, useEffect} from 'react';
import Container from 'react-bootstrap/Container';
import { WorkoutList } from '../../components/workout/WorkoutList';
import {Context as WorkoutContext} from '../../context/WorkoutContext';

export const ViewWorkoutsPage = ()=> 
{
    const {state, fetchWorkouts} = useContext(WorkoutContext);

    useEffect(()=> {
        const processWorkouts = async () => {
            await fetchWorkouts();
        }
       processWorkouts();
    }, []);
    
    return (
        <Container>
            <h3>Your Workouts</h3>
            <WorkoutList workoutList={state} />
        </Container>
    )
}