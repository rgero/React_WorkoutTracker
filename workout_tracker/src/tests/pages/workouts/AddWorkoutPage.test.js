import ShallowRenderer from 'react-test-renderer/shallow'
import React from 'react';
import { AddWorkoutPage } from '../../../pages/workouts/AddWorkoutPage';

let realUseContext;
let useContextMock;

beforeEach(()=> {
    realUseContext = React.useContext;
    useContextMock = React.useContext = jest.fn();
})

afterEach(()=> {
    React.useContext = realUseContext;
})

test('Renders AddWorkoutPage Page and checks against snapshot', ()=> {
    useContextMock.mockReturnValue({
        state: {},
        createWorkout: jest.fn()
    });
    
    const pageRender = new ShallowRenderer().render(
        <AddWorkoutPage />
    );

    expect(pageRender).toMatchSnapshot();
})