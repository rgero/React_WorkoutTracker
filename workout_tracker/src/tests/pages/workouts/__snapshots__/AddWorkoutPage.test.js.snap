// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders AddWorkoutPage Page and checks against snapshot 1`] = `
<Container
  fluid="md"
>
  <h1>
    Create a new workout
  </h1>
  <WorkoutForm
    onSubmit={[Function]}
  />
</Container>
`;
